File: app/src/main/java/com/serenegiant/janusrtcandroid/CallFragment.java
package com.serenegiant.janusrtcandroid;/*
 *  Copyright 2015 The WebRTC Project Authors. All rights reserved.
 *  Modified 2018 by saki t_saki@serenegiant.com
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree. An additional intellectual property rights grant can be found
 *  in the file PATENTS.  All contributing project authors may
 *  be found in the AUTHORS file in the root of the source tree.
 */

import android.app.Activity;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.SeekBar;
import android.widget.TextView;

import org.webrtc.RendererCommon.ScalingType;

import java.util.Locale;

/**
 * Fragment for call control.
 */
public class CallFragment extends Fragment {
	private static final boolean DEBUG = true;	// set false on production
	private static final String TAG = CallFragment.class.getSimpleName();

	private TextView contactView;
	private ImageButton cameraSwitchButton;
	private ImageButton videoScalingButton;
	private ImageButton toggleMuteButton;
	private TextView captureFormatText;
	private SeekBar captureFormatSlider;
	private OnCallEvents callEvents;
	private ScalingType scalingType;
	private boolean videoCallEnabled = true;
	
	/**
	 * Call control interface for container activity.
	 */
	public interface OnCallEvents {
		void onCallHangUp();
		
		void onCameraSwitch();
		
		void onVideoScalingSwitch(ScalingType scalingType);
		
		void onCaptureFormatChange(int width, int height, int framerate);
		
		boolean onToggleMic();
	}
	
	@Override
	public View onCreateView(
		@NonNull final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {

		final View controlView = inflater.inflate(R.layout.fragment_call, container, false);
		
		// Create UI controls.
		contactView = controlView.findViewById(R.id.contact_name_call);
		final ImageButton disconnectButton = controlView.findViewById(R.id.button_call_disconnect);
		cameraSwitchButton = controlView.findViewById(R.id.button_call_switch_camera);
		videoScalingButton = controlView.findViewById(R.id.button_call_scaling_mode);
		toggleMuteButton = controlView.findViewById(R.id.button_call_toggle_mic);
		captureFormatText = controlView.findViewById(R.id.capture_format_text_call);
		captureFormatSlider = controlView.findViewById(R.id.capture_format_slider_call);
		
		// Add buttons click events.
		disconnectButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(final View view) {
				callEvents.onCallHangUp();
			}
		});
		
		cameraSwitchButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(final View view) {
				callEvents.onCameraSwitch();
			}
		});
		
		videoScalingButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(final View view) {
				if (scalingType == ScalingType.SCALE_ASPECT_FILL) {
					videoScalingButton.setBackgroundResource(R.drawable.ic_action_full_screen);
					scalingType = ScalingType.SCALE_ASPECT_FIT;
				} else {
					videoScalingButton.setBackgroundResource(R.drawable.ic_action_return_from_full_screen);
					scalingType = ScalingType.SCALE_ASPECT_FILL;
				}
				callEvents.onVideoScalingSwitch(scalingType);
			}
		});
		scalingType = ScalingType.SCALE_ASPECT_FILL;
		
		toggleMuteButton.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(final View view) {
				boolean enabled = callEvents.onToggleMic();
				toggleMuteButton.setAlpha(enabled ? 1.0f : 0.3f);
			}
		});
		
		return controlView;
	}
	
	@Override
	public void onStart() {
		super.onStart();
		
		boolean captureSliderEnabled = false;
		final Bundle args = getArguments();
		if (args != null) {
			final int roomId = args.getInt(CallActivity.EXTRA_ROOMID);
			contactView.setText(String.format(Locale.US, "%d", roomId));
			videoCallEnabled = args.getBoolean(CallActivity.EXTRA_VIDEO_CALL, true);
			captureSliderEnabled = videoCallEnabled
				&& args.getBoolean(CallActivity.EXTRA_VIDEO_CAPTUREQUALITYSLIDER_ENABLED, false);
		}
		if (!videoCallEnabled) {
			cameraSwitchButton.setVisibility(View.INVISIBLE);
		}
		if (captureSliderEnabled) {
			captureFormatSlider.setOnSeekBarChangeListener(
				new CaptureQualityController(captureFormatText, callEvents));
		} else {
			captureFormatText.setVisibility(View.GONE);
			captureFormatSlider.setVisibility(View.GONE);
		}
	}
	
	// TODO(sakal): Replace with onAttach(Context) once we only support API level 23+.
	@SuppressWarnings("deprecation")
	@Override
	public void onAttach(final Activity activity) {
		super.onAttach(activity);
		callEvents = (OnCallEvents) activity;
	}
}

File: app/src/main/java/com/serenegiant/janusrtcandroid/CallFragment.kt
package com.serenegiant.janusrtcandroid

import android.content.Context
import android.widget.TextView
import android.widget.ImageButton
import android.widget.SeekBar
import org.webrtc.RendererCommon.ScalingType
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.fragment.app.Fragment
import java.util.*

/*
 *  Copyright 2015 The WebRTC Project Authors. All rights reserved.
 *  Modified 2018 - 2022 by saki t_saki@serenegiant.com
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree. An additional intellectual property rights grant can be found
 *  in the file PATENTS.  All contributing project authors may
 *  be found in the AUTHORS file in the root of the source tree.
 */ /**
 * Fragment for call control.
 */
class CallFragment : Fragment() {
	private lateinit var contactView: TextView
	private lateinit var cameraSwitchButton: ImageButton
	private lateinit var videoScalingButton: ImageButton
	private lateinit var toggleMuteButton: ImageButton
	private lateinit var captureFormatText: TextView
	private lateinit var captureFormatSlider: SeekBar
	private var scalingType: ScalingType? = null
	private var videoCallEnabled = true

	/**
	 * Call control interface for container activity.
	 */
	interface OnCallEvents {
		fun onCallHangUp()
		fun onCameraSwitch()
		fun onVideoScalingSwitch(scalingType: ScalingType?)
		fun onCaptureFormatChange(width: Int, height: Int, framerate: Int)
		fun onToggleMic(): Boolean
	}

	// TODO(sakal): Replace with onAttach(Context) once we only support API level 23+.
	override fun onAttach(context: Context) {
		super.onAttach(context)
		if (DEBUG) Log.v(TAG, "onAttach:")
	}

	override fun onCreateView(
		inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
		if (DEBUG) Log.v(TAG, "onCreateView:")

		val controlView = inflater.inflate(R.layout.fragment_call, container, false)
		// Create UI controls.
		contactView = controlView.findViewById(R.id.contact_name_call)
		val disconnectButton = controlView.findViewById<ImageButton>(R.id.button_call_disconnect)
		cameraSwitchButton = controlView.findViewById(R.id.button_call_switch_camera)
		videoScalingButton = controlView.findViewById(R.id.button_call_scaling_mode)
		toggleMuteButton = controlView.findViewById(R.id.button_call_toggle_mic)
		captureFormatText = controlView.findViewById(R.id.capture_format_text_call)
		captureFormatSlider = controlView.findViewById(R.id.capture_format_slider_call)

		val callEvents = requireOnCallEvents()
		// Add buttons click events.
		disconnectButton.setOnClickListener { callEvents.onCallHangUp() }
		cameraSwitchButton.setOnClickListener { callEvents.onCameraSwitch() }
		videoScalingButton.setOnClickListener {
			scalingType = if (scalingType == ScalingType.SCALE_ASPECT_FILL) {
				videoScalingButton.setBackgroundResource(R.drawable.ic_action_full_screen)
				ScalingType.SCALE_ASPECT_FIT
			} else {
				videoScalingButton.setBackgroundResource(R.drawable.ic_action_return_from_full_screen)
				ScalingType.SCALE_ASPECT_FILL
			}
			callEvents.onVideoScalingSwitch(scalingType)
		}
		scalingType = ScalingType.SCALE_ASPECT_FILL
		toggleMuteButton.setOnClickListener {
			val enabled = callEvents.onToggleMic()
			toggleMuteButton.alpha = if (enabled) 1.0f else 0.3f
		}
		return controlView
	}

	override fun onStart() {
		super.onStart()
		if (DEBUG) Log.v(TAG, "onStart:")
		var captureSliderEnabled = false
		val args = arguments
		if (args != null) {
			val roomId = args.getLong(CallActivity.EXTRA_ROOMID, 0)
			contactView.text = String.format(Locale.US, "%d", roomId)
			videoCallEnabled = args.getBoolean(CallActivity.EXTRA_VIDEO_CALL, true)
			captureSliderEnabled = (videoCallEnabled
				&& args.getBoolean(CallActivity.EXTRA_VIDEO_CAPTUREQUALITYSLIDER_ENABLED, false))
		}
		if (!videoCallEnabled) {
			cameraSwitchButton.visibility = View.INVISIBLE
		}
		if (captureSliderEnabled) {
			captureFormatSlider.setOnSeekBarChangeListener(
				CaptureQualityController(captureFormatText, requireOnCallEvents())
			)
		} else {
			captureFormatText.visibility = View.GONE
			captureFormatSlider.visibility = View.GONE
		}
	}

	override fun onDetach() {
		if (DEBUG) Log.v(TAG, "onDetach:")
		super.onDetach()
	}

	private fun requireOnCallEvents(): OnCallEvents {
		return requireContext() as OnCallEvents
	}

	companion object {
		private const val DEBUG = true // set false on production
		private val TAG = CallFragment::class.java.simpleName
	}
}
